## Introduction
We choose option Neural FCA. We choose following tabular datasets and targets

| **№** | **Dataset name**                                                                      | **Description**                        | **Size**     | **Target**     |
|----------|--------------------------------------------------------------------------------------------|---------------------------------------------|-------------------|---------------------|
| 1        | [Palmer penguins](https://archive.ics.uci.edu/dataset/690/palmer+penguins-3)         | An introductory dataset presented as an <br>alternative to Iris and useful for teaching data <br>exploration/visualization. Data comes from 3 penguin  <br> species in the islands of Palmer Archipelago, Antarctica.   | (344, 9)    | species<br>(3 possible values)             |                                                         |   | ~                 | ~                   |
| 2        | [Credit scoring](https://www.kaggle.com/competitions/fintech-credit-scoring/overview) | Bank applicants’ personal data and the fact<br> of default dataset. | (181000, 19) | default flag<br>(2 possible values)       |
| 3        | [Zoo](https://archive.ics.uci.edu/dataset/111/zoo)                                    | A simple database containing 17 Boolean-valued<br>attributes of 101 different animals.    | (101, 17)    | type <br>(7 possible values)             |
| 4        | [The Complete Pokemon Dataset](https://www.kaggle.com/datasets/rounakbanik/pokemon)                   | Dataset of all 802 Pokemon from all <br>Seven Generations of Pokemon.      | (802,41)                                                                             | legendary flag<br>(2 possible values)                          |

Motivation behind our choice is following. **Palmer penguins** is well known benchmark dataset, which is useful for comparisons of different models. **Credit scoring** is close to real world medium sized dataset, which is useful to for scalability checks and performance checks on noisy data. **Zoo** describes structure of real animals relations, which is similar to concept description. In contrast, **The Complete Pokemon Dataset** describes imaginary creatures relations. It is interesting to compare real world system of relations with man made system of relations.

## Preprocessing
We binarize all datasets’ features and targets. The
description of preprocessing for every dataset is in the following table

| № | Dataset name | Preprocessing Description | Final size | Target |
|---|--------------|---------------------------|------------|--------|
| 1 | [Palmer penguins (PP)](https://archive.ics.uci.edu/dataset/690/palmer+penguins-3) | We drop 'rowid' column. We drop rows '3' and '271' because all features of the rows are NaN.<br>We use ohe encoding on categorical features ['island','year','sex'].<br>We discretize and perform ohe encoding with continuous features ['bill\_length\_mm', 'bill\_depth\_mm', 'flipper\_length\_mm', 'body\_mass\_g'].<br>We use features ['bill\_length\_mm', 'bill\_depth\_mm','flipper\_length\_mm', 'body\_mass\_g'].<br>We use quantile-based discretization function.<br>We divide values into 3 quantiles.<br>We impute missing values with KNNimputer. | (342, 19) | Chinstrap species flag |
|   |   |   |   |   |
| 2 | [Credit scoring (C)](https://www.kaggle.com/competitions/fintech-credit-scoring/overview) | Because of the small number of missing values and the big size of the dataset, we drop all rows with missing values.<br>Because our computation capabilities are limited, we leave the first 90,000 rows.<br>We apply ohe encoding on categorical features ['good\_work\_flg','Air\_flg','car\_type\_flg','car\_own\_flg','gender\_cd','home\_address\_cd','work\_address\_cd','SNA','first\_time\_cd','education\_cd'].<br>We discretize and perform ohe encoding with continuous features ['region\_rating', 'appl\_rej\_cnt','out\_request\_cnt','age','income','Score\_bki'].<br>We use quantile-based discretization function.<br>We divide values into 3 quantiles. | (90,000, 45) | Default flag |
|   |   |   |   |   |
| 3 | [Zoo (Z)](https://archive.ics.uci.edu/dataset/111/zoo) | All features except 'legs' are binary.<br>We perform ohe encoding on 'legs' feature. | (101, 21) | Type 0 or 1 flag |
|   |   |   |   |   |
| 4 | [The Complete Pokemon Dataset (P)](https://www.kaggle.com/datasets/rounakbanik/pokemon) | Since the dataset is very wide, we select only 'agains\_...' features and some generic like 'hp' and others.<br>We discretize and perform ohe encoding on all features.<br>We use quantile-based discretization function.<br>We divide all 'againts\_...' features and generic features into 3 and 2 quantiles respectively. | (801, 50) | Legendary flag |


## ML

First, using 'lazypredict' library we evaluated 30 classification models (see Tables 6,7,8 and 9). We see that there are good performing models that are not in the proposed list. For this reason we add 6 more models for classification.  

We tune models' parameters using cross-validation (5 fold). Evaluation results of best models are presented in Table 10. In Table 11 we aggregated the results in pivot table. We see that all models have high scores for classical datasets (PP and Z) and struggle with close to real data (C and P). It is interesting to note that NearestCentroid model has highest score for C, but lower scores in all other datasets; LinearDiscriminantAnalysis model on average has highest among all datasets.

Source code of preprocessing, models' fit and  evaluation is in following Jupyter notebook [hse.kamran.uz/osda23/fca/osda23\_1.ipynb](https://hse.kamran.uz/osda23/fca/osda23_1.ipynb). All computations were performed on Apple M1 Pro, 32 GB.

Evaluation results of best models of the cross-validation (5 fold) models’ parameter tuning. In bold best model’s highest F1 score

<img src="https://hse.kamran.uz/osda23/fca/hw2_9.png" alt="drawing" width="400"/>


Pivot table of every model and dataset for F1 score.
In bold highest F1 score. Second highest F1 score is underlined

<img src="https://hse.kamran.uz/osda23/fca/hw2_10.png" alt="drawing" width="400"/>


## NeuralFCA Application

In [1]:
#@title install libs

!pip install fcapy[all]
!pip install frozendict
!pip install ipynb
!pip install sparselinear
!pip install bitsets
!pip install bitarray
import torch
!pip install torch-scatter -f https://data.pyg.org/whl/torch-2.0.0+cuda118.html
!pip install torch-sparse -f https://data.pyg.org/whl/torch-2.0.0+cuda118.html
!pip install torch-cluster -f https://data.pyg.org/whl/torch-2.0.0+cuda118.html
!pip install git+https://github.com/pyg-team/pytorch_geometric.git

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 162.9/162.9 kB 1.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 118.8/118.8 kB 7.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 288.7/288.7 kB 10.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.6/43.6 kB 2.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 25.1 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 49.0 MB/s eta 0:00:00
  Created wheel for wget: filename=wget-3.2-py3-none-any.whl size=9655 sha256=af8879dfcf5e13e38d18a8b1e573289bdb81333e24ffdf1e5c73e7cab39a79bb
  Stored in directory: /root/.cache/pip/wheels/8b/f1/7f/5c94f0a7a505ca1c81cd1d9208ae2064675d97582078e6c769
Successfully built wget


Looking in links: https://data.pyg.org/whl/torch-2.0.0+cuda118.html
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 108.0/108.0 kB 2.3 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for torch-scatter: filename=torch_scatter-2.1.2-cp310-cp310-linux_x86_64.whl size=495091 sha256=576a550091c49383c5fe5498956bf31bf4c65cd053453259f401c853f1b8097a
  Stored in directory: /root/.cache/pip/wheels/92/f1/2b/3b46d54b134259f58c8363568569053248040859b1a145b3ce
Successfully built torch-scatter
Looking in links: https://data.pyg.org/whl/torch-2.0.0+cuda118.html
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 210.0/210.0 kB 2.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for torch-sparse: filename=torch_sparse-0.6.18-cp310-cp310-linux_x86_64.whl size=1035675 sha256=72e2fe703ea19897333d30c6e34d797bfaeff0f188808345f69c23dee2467bda
  Stored in directory: /root/.cache/pip/wheels/c9/dd/0f/a6a16f9f3b0236733d257b4b4ea91b548b984a341ed3b8f38c
Successfully buil

In [114]:
# @title neural_lib.py
%%writefile neural_lib.py

from dataclasses import dataclass
from typing import List, Tuple, FrozenSet, Set, Dict
import pandas as pd

from fcapy.lattice import ConceptLattice
from fcapy.lattice.formal_concept import FormalConcept
from fcapy.poset import POSet
from fcapy.visualizer.line_layouts import calc_levels

import torch
from sparselinear import SparseLinear


@dataclass(eq=False)
class DisjunctiveNeuron:
    intent: FrozenSet[str]
    level: int

    def __eq__(self, other: 'DisjunctiveNeuron'):
        return self.intent == other.intent and self.level == other.level

    def __lt__(self, other: 'DisjunctiveNeuron'):
        return self.intent & other.intent == other.intent and self.level > other.level

    def __le__(self, other: 'DisjunctiveNeuron'):
        return self < other or self == other

    def __hash__(self):
        return hash((self.intent, self.level))


class ConceptNetwork:
    def __init__(self, poset: POSet, network=None, attributes: Tuple[str] = None, targets: Tuple[str] = None):
        self._poset = poset
        self._network = network
        self._attributes = attributes
        self._targets = targets

    @property
    def poset(self) -> POSet:
        return self._poset

    @property
    def network(self) -> torch.nn.Sequential:
        return self._network

    @property
    def attributes(self) -> Tuple[str]:
        return self._attributes

    @property
    def targets(self):
        return self._targets

    def trace_description(self, description: FrozenSet[str], include_targets: bool = False) -> Set[int]:
        P = self.poset

        tops_activated = [node for node in P.tops if P[node].intent & description == P[node].intent]
        activated_nodes = set(tops_activated)
        for node in tops_activated:
            activated_nodes |= P.descendants(node)
        if not include_targets:
            activated_nodes -= set(P.bottoms)

        return activated_nodes

    @classmethod
    def from_lattice(
            cls,
            lattice: ConceptLattice, best_concepts_indices: List[int],
            targets: Tuple[str]
    ):
        assert lattice.is_monotone, 'The lattice should be monotone'

        targets = tuple(targets)

        attrs_tpl = tuple(lattice[lattice.bottom].intent)
        P = cls._poset_from_best_concepts(lattice[best_concepts_indices], targets, attrs_tpl)
        P = cls._fill_levels(P)
        return cls(P, None, attributes=attrs_tpl, targets=targets)

    def fit(
            self,
            X_df: 'pd.DataFrame[bool]', y: 'pd.Series[bool]',
            loss_fn=torch.nn.CrossEntropyLoss(), nonlinearity=torch.nn.ReLU,
            n_epochs: int = 2000
    ):
        X = torch.tensor(X_df[list(self.attributes)].values).float()
        y = torch.tensor(y.values).long()

        self._network = self._poset_to_network(self.poset, nonlinearity)

        optimizer = torch.optim.Adam(self.network.parameters())

        for t in range(n_epochs):
            optimizer.zero_grad()
            y_pred = self.network(X)
            loss = loss_fn(y_pred, y)
            loss.backward()
            optimizer.step()

    def predict_proba(self, X_df: 'pd.DataFrame[bool]') -> torch.Tensor:
        X = torch.tensor(X_df[list(self.attributes)].values).float()
        return self.network(X)

    def predict(self, X_df: 'pd.DataFrame[bool]') -> torch.Tensor:
        return self.predict_proba(X_df).argmax(1)

    def edge_weights_from_network(self) -> Dict[Tuple[int, int], float]:
        max_level = self.poset[self.poset.bottoms[0]].level
        nodes_per_levels = {lvl: [] for lvl in range(max_level + 1)}
        for node_i, node in enumerate(self.poset):
            nodes_per_levels[node.level].append(node_i)
        nodes_per_levels = [nodes_per_levels[lvl] for lvl in range(max_level + 1)]

        edge_weights = {}
        for layer_i, nodes in enumerate(nodes_per_levels[:-1]):
            next_nodes = nodes_per_levels[layer_i+1]

            nn_layer = self.network[layer_i*2]
            idxs = nn_layer.weight.indices().numpy().T.tolist()
            vals = nn_layer.weight.values().numpy()

            for (child_i, parent_i), v in zip(idxs, vals):
                edge_weights[(nodes[parent_i], next_nodes[child_i])] = v
        return edge_weights

    @staticmethod
    def _poset_from_best_concepts(
            best_concepts: List[FormalConcept], targets: Tuple[str], attrs_tpl: Tuple[str]
    ) -> POSet:
        P_best = POSet(best_concepts)
        lvls = calc_levels(P_best)[0]
        lvls = [lvl + 1 for lvl in lvls]
        target_lvl = max(lvls) + 1

        attrs_set = set(attrs_tpl)

        best_neurons = [DisjunctiveNeuron(frozenset(c.intent), lvl) for c, lvl in zip(P_best, lvls)]
        first_level_neurons = [DisjunctiveNeuron(frozenset({m}), 0) for m in attrs_tpl]
        last_level_neurons = [DisjunctiveNeuron(frozenset({f"y={y}"} | attrs_set), target_lvl) for y in targets]
        return POSet(first_level_neurons + best_neurons + last_level_neurons)

    @staticmethod
    def _fill_levels(poset: POSet) -> POSet:
        nodes_i = sorted(range(len(poset)), key=lambda node_i: poset[node_i].level)
        for node_i in nodes_i:
            children_i = poset.children(node_i)
            if len(children_i) == 0:
                continue

            max_children_level = max([poset[child_i].level for child_i in children_i])
            for lvl in range(poset[node_i].level+1, max_children_level):
                poset.add(DisjunctiveNeuron(poset[node_i].intent, lvl))
        return poset

    @staticmethod
    def _poset_to_network(poset: POSet, nonlinearity: type = torch.nn.ReLU) -> 'torch.nn.Sequential':
        max_level = poset[poset.bottoms[0]].level
        nodes_per_levels = {lvl: [] for lvl in range(max_level + 1)}
        for node_i, node in enumerate(poset):
            nodes_per_levels[node.level].append(node_i)
        nodes_per_levels = [nodes_per_levels[lvl] for lvl in range(max_level + 1)]

        connectivities = []
        for layer_i, layer in enumerate(nodes_per_levels[1:]):
            layer_i += 1
            prev_layer = nodes_per_levels[layer_i - 1]
            layer_con = [(layer.index(node), prev_layer.index(parent))
                         for node in layer for parent in poset.parents(node)]
            connectivities.append(layer_con)

        linear_layers = []
        for layer_i in range(max_level):
            con = torch.tensor(connectivities[layer_i]).T
            layer = SparseLinear(len(nodes_per_levels[layer_i]), len(nodes_per_levels[layer_i + 1]), connectivity=con)
            linear_layers.append(layer)

        layers = [layer for ll in linear_layers for layer in [ll, nonlinearity()]][:-1] + [torch.nn.Softmax(dim=1)]
        model_sparse = torch.nn.Sequential(*layers)
        return model_sparse


def neuron_label_func(el_i: int, P: POSet, M: set, only_new_attrs: bool = True):
    el = P[el_i]

    if len(el.intent - M) > 0:  # if target node
        attrs_to_show = list(el.intent - M)
    else:
        attrs_to_show = set(el.intent)
        if only_new_attrs:
            for parent_i in P.parents(el_i):
                attrs_to_show = attrs_to_show - P[parent_i].intent

        attrs_to_show = list(attrs_to_show)
    return ','.join(attrs_to_show)

Overwriting neural_lib.py


In [134]:
#@title import libs

import numpy as np
import pandas as pd
import requests
from io import StringIO
from tqdm import tqdm

from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score, precision_score, jaccard_score, recall_score, accuracy_score, classification_report

from fcapy.context import FormalContext
from fcapy.lattice import ConceptLattice

from fcapy.visualizer import LineVizNx
import neural_lib as nl
from fcapy.utils.utils import powerset
import matplotlib.pyplot as plt

plt.rcParams['figure.facecolor'] = (1,1,1,1)


from fcapy import LIB_INSTALLED
if LIB_INSTALLED['numpy']:
    import numpy as np

from sparselinear import SparseLinear

In [35]:
#@title import data

def get(path):
    # Using the requests library to handle the URL
    try:
        response = requests.get(path)
        response.raise_for_status()  # This will raise an HTTPError if the HTTP request returned an unsuccessful status code.

        # Reading the content of the file into a pandas DataFrame
        df = pd.read_csv(StringIO(response.text)).astype(bool)

    except requests.exceptions.HTTPError as errh:
        print("Http Error:", errh)
    except requests.exceptions.ConnectionError as errc:
        print("Error Connecting:", errc)
    except requests.exceptions.Timeout as errt:
        print("Timeout Error:", errt)
    except requests.exceptions.RequestException as err:
        print("Oops: Something Else", err)
    except pd.errors.EmptyDataError:
        print("No data: Empty Data Received")
    except Exception as e:
        print("An error occurred:", e)
    return df

root = 'https://hse.kamran.uz/osda23/fca/'

# root = ''

pp = get(f'{root}prepared_pp.csv')
c  = get(f'{root}prepared_c.csv')
z  = get(f'{root}prepared_z.csv')
p  = get(f'{root}prepared_p.csv')


y_pp = get(f'{root}target_pp.csv')
y_c  = get(f'{root}target_c.csv')
y_z  = get(f'{root}target_z.csv')
y_p  = get(f'{root}target_p.csv')

In [350]:
#@title train | f1_score

# DELTA_STABILITY
from numpy.random import seed as np_seed
import tensorflow as tf
import random
import math

RANDOM_SEED = 42
np_seed(RANDOM_SEED)
random.seed(RANDOM_SEED)
tf.random.set_seed(RANDOM_SEED)


def log_stability_lbound(c_i, lattice: ConceptLattice, n_bin_attrs: int) -> float:
    extent_i = set(lattice[c_i].extent_i)
    children_i = lattice.children(c_i)
    if children_i:
        bound = min(len(extent_i - set(lattice[child_i].extent_i)) for child_i in children_i)
    else:
        bound = math.inf
    bound -= math.log2(n_bin_attrs)
    return bound

def delta_stability(c_i, lattice: ConceptLattice, n_bin_attrs: int) -> float:
    return log_stability_lbound(c_i, lattice, n_bin_attrs)+math.log2(n_bin_attrs)



def get_train_test(X, y):
    X.index = X.index.astype('str')
    y.index = y.index.astype('str')
    X_train, X_test, y_train, y_test = train_test_split(X, y,
                                                        test_size=0.1,
                                                        random_state=RANDOM_SEED,
                                                        shuffle=True,
                                                        stratify=y)
    return X_train, X_test, y_train, y_test

X = [
    pp,
     z,
     p,
     c,
     ]
y = [
    y_pp,
     y_z,
     y_p,
     y_c,
     ]
models = []
for X_, y_ in tqdm(zip(X,y)):
    m = []
    # Split the data to train and test
    X_train, X_test, y_train, y_test = get_train_test(X_, y_)

    # Put binarized data in FormalContext and compute monotone ConceptLattice
    K_train = FormalContext(data = X_train.values, target=y_train.values, attribute_names=X_train.columns)
    L = ConceptLattice.from_context(K_train, algo='Sofia', is_monotone=True)

    # Compute F1 score for each formal concept (assuming that an object is predicted True if it is in the extent of the concept)
    for i in range(len(L)):
        y_preds = np.zeros(K_train.n_objects)
        y_preds[list(L[i].extent_i)] = 1
        L[i].measures['f1_score']=f1_score(y_train, y_preds)

    # Select indices of the best concepts from the lattice
    best_concepts = list(L.measures['f1_score'].argsort()[::-1])
    for i in range(len(best_concepts)):
        if len({g_i for c in L[best_concepts[:i]] for g_i in c.extent_i})==K_train.n_objects:
            print()
            print('Number of best concepts to cover all objects: ',i)
            best_concepts = best_concepts[:i]
            break

    # Construct neural network based on concept lattice
    cn = nl.ConceptNetwork.from_lattice(L, best_concepts, sorted(set(y_train.iloc[:,0])))
    cn.fit(X_train, y_train.iloc[:,0],  n_epochs =1000)
    y_pred = cn.predict(X_test).numpy()
    print()
    print('Recall score:', recall_score(y_test.values.astype('int'), y_pred))
    print('F1     score:', f1_score(y_test.values.astype('int'), y_pred))
    print('Accuracy score:', accuracy_score(y_test.values.astype('int'), y_pred))
    print()
    models.append([cn, K_train, L, [X_train, X_test, y_train, y_test]])


0it [00:00, ?it/s]


Number of best concepts to cover all objects:  8


1it [00:03,  3.71s/it]


Recall score: 0.0
F1     score: 0.0
Accuracy score: 0.8


Number of best concepts to cover all objects:  13


2it [00:08,  4.31s/it]


Recall score: 1.0
F1     score: 0.7777777777777778
Accuracy score: 0.6363636363636364


Number of best concepts to cover all objects:  13


3it [00:14,  5.01s/it]


Recall score: 0.0
F1     score: 0.0
Accuracy score: 0.9135802469135802


Number of best concepts to cover all objects:  21


4it [07:20, 110.06s/it]


Recall score: 0.0
F1     score: 0.0
Accuracy score: 0.8667777777777778



In [349]:
#@title train | delta_stability

# DELTA_STABILITY
from numpy.random import seed as np_seed
import tensorflow as tf
import random
import math

RANDOM_SEED = 42
np_seed(RANDOM_SEED)
random.seed(RANDOM_SEED)
tf.random.set_seed(RANDOM_SEED)


def log_stability_lbound(c_i, lattice: ConceptLattice, n_bin_attrs: int) -> float:
    extent_i = set(lattice[c_i].extent_i)
    children_i = lattice.children(c_i)
    if children_i:
        bound = min(len(extent_i - set(lattice[child_i].extent_i)) for child_i in children_i)
    else:
        bound = math.inf
    bound -= math.log2(n_bin_attrs)
    return bound

def delta_stability(c_i, lattice: ConceptLattice, n_bin_attrs: int) -> float:
    return log_stability_lbound(c_i, lattice, n_bin_attrs)+math.log2(n_bin_attrs)



def get_train_test(X, y):
    X.index = X.index.astype('str')
    y.index = y.index.astype('str')
    X_train, X_test, y_train, y_test = train_test_split(X, y,
                                                        test_size=0.1,
                                                        random_state=RANDOM_SEED,
                                                        shuffle=True,
                                                        stratify=y)
    return X_train, X_test, y_train, y_test

X = [
    pp,
     z,
     p,
     c,
     ]
y = [
    y_pp,
     y_z,
     y_p,
     y_c,
     ]
models = []
for X_, y_ in tqdm(zip(X,y)):
    m = []
    # Split the data to train and test
    X_train, X_test, y_train, y_test = get_train_test(X_, y_)

    # Put binarized data in FormalContext and compute monotone ConceptLattice
    K_train = FormalContext(data = X_train.values, target=y_train.values, attribute_names=X_train.columns)
    L = ConceptLattice.from_context(K_train, algo='Sofia', is_monotone=True)

    # Compute F1 score for each formal concept (assuming that an object is predicted True if it is in the extent of the concept)
    for i in range(len(L)):
        y_preds = np.zeros(K_train.n_objects)
        y_preds[list(L[i].extent_i)] = 1
        L[i].measures['delta_stability'] = delta_stability(i, L, X_train.shape[1])

    # Select indices of the best concepts from the lattice
    best_concepts = list(L.measures['delta_stability'].argsort()[::-1])
    for i in range(len(best_concepts)):
        if len({g_i for c in L[best_concepts[:i]] for g_i in c.extent_i})==K_train.n_objects:
            print()
            print('Number of best concepts to cover all objects: ',i)
            best_concepts = best_concepts[:i]
            break

    # Construct neural network based on concept lattice
    cn = nl.ConceptNetwork.from_lattice(L, best_concepts, sorted(set(y_train.iloc[:,0])))
    cn.fit(X_train, y_train.iloc[:,0],  n_epochs =1000)
    y_pred = cn.predict(X_test).numpy()
    print()
    print('Recall score:', recall_score(y_test.values.astype('int'), y_pred))
    print('F1     score:', f1_score(y_test.values.astype('int'), y_pred))
    print('Accuracy score:', accuracy_score(y_test.values.astype('int'), y_pred))
    print()
    models.append([cn, K_train, L, [X_train, X_test, y_train, y_test]])


0it [00:00, ?it/s]


Number of best concepts to cover all objects:  1


1it [00:01,  1.88s/it]


Recall score: 0.8571428571428571
F1     score: 0.923076923076923
Accuracy score: 0.9714285714285714


Number of best concepts to cover all objects:  1


2it [00:03,  1.69s/it]


Recall score: 1.0
F1     score: 1.0
Accuracy score: 1.0


Number of best concepts to cover all objects:  1


3it [00:05,  1.98s/it]


Recall score: 0.0
F1     score: 0.0
Accuracy score: 0.9135802469135802


Number of best concepts to cover all objects:  1


4it [01:47, 26.96s/it]


Recall score: 0.0
F1     score: 0.0
Accuracy score: 0.8667777777777778



## Conclusion

We applied NeuralFCA to 4 differnet datasets. We used F1 score and Delta stability metrics to for finding best concepts for ConceptNetwork learning. We see that version with delta stability gives better results

For difficult datasets P and C we got unsatisfactory results (F1 scores 0 and 0 vs 0.33 and 0.66 in baseline respectivelly). We believe that this is consequence of inherit inconsistency of manmade system of Pokemons and chaotic nature of credit defaults

We got high results for PP and Z datasets (F1 scores 0.92 and 1 vs 1 and 1 in basesline respectively). We believe that this is consequence of systemic nature of the wild life features